devito/at_controller.py

```diff
@@ -0,0 +1,315 @@
+        self._at_work_dir = os.path.join(isat_dir_path, "Auto-tuning")  # working dir for at
+
+        at_file_dir = os.path.join(self._at_work_dir, "Isat-files")  # dir for keeping files necessary for at
+
+        # all these files are necessary for auto tuning process and they have to be in same dir as tuned file
+        # do not change these names
+        self._isat_build_f = "isat-build"
+        self._isat_clean_f = "isat-clean"
+        self._isat_test_f = "isat-test"
+        self._make_f = "Makefile"
+
+        self._build_f_path = os.path.join(at_file_dir, self._isat_build_f)
+        self._clean_f_path = os.path.join(at_file_dir, self._isat_clean_f)
+        self._test_f_path = os.path.join(at_file_dir, self._isat_test_f)
+        self._make_f_path = os.path.join(at_file_dir, self._make_f)
+
+        self._at_parent_folder = ""  # saves the location of current at parent folder
+        self._at_src_dir = "src"  # source folder where file which we'll auto tune will be kept
+        self._at_dst_dir = "dst"  # destination folder where at result will be kept
+
+        # for simplicity renaming all the tuned files to this name
+        # their parent dir will contain the time and space order in cpp file kernel
+        self._at_f_name = "tune_me"
+        shell_ref_str = "#!/bin/sh"
+
+        if not os.path.exists(self._at_work_dir):  # creates main working directory
+            os.mkdir(self._at_work_dir)
+
+        if os.path.exists(at_file_dir):  # making sure that all files are up to date
+            shutil.rmtree(at_file_dir)
+
+        os.mkdir(at_file_dir)
+
+        with open(self._build_f_path, 'w') as f:  # creates isat-build
+            f.writelines([shell_ref_str, "\nmake build"])
+        set_x_permission(self._build_f_path)
+
+        with open(self._clean_f_path, 'w') as f:  # create isat-clean
+            f.writelines([shell_ref_str, "\nmake clean"])
+        set_x_permission(self._clean_f_path)
+
+        with open(self._test_f_path, 'w') as f:  # create isat-test
+            f.writelines([shell_ref_str, "\n./%s.out" % self._at_f_name])
+        set_x_permission(self._test_f_path)
+
+        with open(self._make_f_path, 'w') as f:  # creates make-file
+            # copied make file contents from one of examples in isat folder. If there's a need feel free to change
+            make_file_contents = ["CXX = %s" % compiler.cc,                 # compiler and its options
+                                  "\nCXX_OPTS = %s" % " ".join(compiler.cflags),
+                                  "\nSRCS = %s.cpp" % self._at_f_name, "\nOUT = %s.out" % self._at_f_name,
+                                  "\n$(OUT): $(SRCS)", "\n\t$(CXX) $(CXX_OPTS) $< -o $@",
+                                  "\nclean:", "\n\t-rm *.out", "\nbuild: $(OUT)"]
+
+            f.writelines(make_file_contents)
+            set_x_permission(self._make_f_path)
+
+    def auto_tune(self, file_path, time_order, space_border):
```

Comment:
Does the auto-tuning really need these "intimate" details of the kernel? Ideally nothing beyond the symbolics should care about the mathematical properties of the kernel.